PIVX-Labs/MyPIVXWallet
# Improve i18n utility, add English fallbacks

## Abstract

This PR adds a well-needed cleanup to the way devs interface with the i18n system.
- We now have a `tr()` utility function, for in-line translations of strings with variables.
- The i18n system now uses the English file as a fallback if a key is missing.
- Translation switching is now a per-key process, rather than entire file.

This means a few small changes to how we previously used it, for ex:

Translating variables in-line is now super easy, i.e:
```js
// i18n key `hello` is "Hello {user}!"
// Which becomes "Hello Bob!"
const user = 'Bob';
const str = tr(translation.hello, [{ user: user }]);
```

Alerts no longer have the additional `variables` param, devs are now expected to use `tr()` in-line to translate alerts on-the-fly.

**Before**
```js
// With variable
createAlert('success', ALERTS.SUCCESSFUL_TX, [{ txid: txid }], 5000);
// Without variable
createAlert('success', ALERTS.SUCCESSFUL_TX, [], 5000);
```

**After**
```js
// With variable
createAlert('success', tr(ALERTS.SUCCESSFUL_TX, [{ txid: txid }]), 5000);
// Without variable
createAlert('success', tr(ALERTS.SUCCESSFUL_TX), 5000);
```

Previously, dynamic JS-selected translations would return `''` or `undefined` to the user in the GUI, if the i18n keys were missing or unset for their language; with the English fallback, this is no longer the case, and users will never be presented with undefined or empty text now.

---

## Testing
To test this PR, it's suggested to attempt these user flows, or variations of these:
- Change your language (multiple times across a few testing sessions and wallets).
- Import or Create a wallet.
- Test various parts of the wallet, ensuring you see no suspicious Notifications or Text which is either empty, or says `undefined`. (Send TXs, try causing errors, change settings, create a proposal on Testnet, etc).

If any errors are found, the PR works unexpectedly, or you have viable suggestions to improve the UX or functionality of the PR, let me know!

---
